src/ancestry_proof.rs

```diff
@@ -0,0 +1,342 @@
+                (pos, leaf)
+            })
+            .collect();
+        self.verify(root, leaves)
+    }
+}
+
+fn calculate_peak_root<
+    'a,
+    T: 'a + PartialEq,
+    M: Merge<Item = T>,
+    // I: Iterator<Item = &'a T>
+>(
+    nodes: Vec<(u64, T)>,
+    peak_pos: u64,
+    // proof_iter: &mut I,
+) -> Result<T> {
+    debug_assert!(!nodes.is_empty(), "can't be empty");
+    // (position, hash, height)
+
+    let mut queue: VecDeque<_> = nodes
+        .into_iter()
+        .map(|(pos, item)| (pos, item, pos_height_in_tree(pos)))
+        .collect();
+
+    let mut sibs_processed_from_back = Vec::new();
+
+    // calculate tree root from each items
+    while let Some((pos, item, height)) = queue.pop_front() {
+        if pos == peak_pos {
+            if queue.is_empty() {
+                // return root once queue is consumed
+                return Ok(item);
+            }
+            if queue
+                .iter()
+                .any(|entry| entry.0 == peak_pos && entry.1 != item)
+            {
+                return Err(Error::CorruptedProof);
+            }
+            if queue
+                .iter()
+                .all(|entry| entry.0 == peak_pos && &entry.1 == &item && entry.2 == height)
+            {
+                // return root if remaining queue consists only of duplicate root entries
+                return Ok(item);
+            }
+            // if queue not empty, push peak back to the end
+            queue.push_back((pos, item, height));
+            continue;
+        }
+        // calculate sibling
+        let next_height = pos_height_in_tree(pos + 1);
+        let (parent_pos, parent_item) = {
+            let sibling_offset = sibling_offset(height);
+            if next_height > height {
+                // implies pos is right sibling
+                let (sib_pos, parent_pos) = (pos - sibling_offset, pos + 1);
+                let parent_item = if Some(&sib_pos) == queue.front().map(|(pos, _, _)| pos) {
+                    let sibling_item = queue.pop_front().map(|(_, item, _)| item).unwrap();
+                    M::merge(&sibling_item, &item)?
+                } else if Some(&sib_pos) == queue.back().map(|(pos, _, _)| pos) {
+                    let sibling_item = queue.pop_back().map(|(_, item, _)| item).unwrap();
+                    M::merge(&sibling_item, &item)?
+                }
+                // handle special if next queue item is descendant of sibling
+                else if let Some(&(front_pos, ..)) = queue.front() {
+                    if height > 0 && is_descendant_pos(sib_pos, front_pos) {
+                        queue.push_back((pos, item, height));
+                        continue;
+                    } else {
+                        return Err(Error::CorruptedProof);
+                    }
+                } else {
+                    return Err(Error::CorruptedProof);
+                };
```

Comment:
Is this needed ? Do you have an example input where this branch would be reached ?